hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/TinyLfuBlockCache.java

```diff
@@ -171,8 +177,10 @@ public Cacheable getBlock(BlockCacheKey cacheKey,
         if ((value != null) && caching) {
           if ((value instanceof HFileBlock) && ((HFileBlock) value).isSharedMem()) {
             value = HFileBlock.deepCloneOnHeap((HFileBlock) value);
+            cacheBlockUtil(cacheKey, value, true);
```

Comment:
I think u can simplify this. No such if else is needed.
Can avoid deepCloneOnHeap and isSharedMem check here. From here call cacheBlock only
In cacheBlock asReferencedHeapBlock() call is newly added and there we have  deepClone as needed.

Reply:
Without this if/else, we would perform `deepCloneOnHeap()` twice right? Just above this line, we do deep clone and then if we just call `cacheBlock()` only, it will again perform `deepClonedOnHeap` because the same condition as above holds true i.e 
```
if ((value instanceof HFileBlock) && ((HFileBlock) value).isSharedMem())
```

Reply:
Pls refer the code in LRUBlockCache.  U can do the deepclone in asReferencedHeapBlock() only based on isSharedMem right?  retain() call is anyways needed

Reply:
> U can do the deepclone in asReferencedHeapBlock() only based on isSharedMem right? retain() call is anyways needed

LRUBlockCache does not perform block.retain() if block is cloned:
```
   * 1. if cache the cloned heap block, its refCnt is an totally new one, it's easy to handle; <br>
   * 2. if cache the original heap block, we're sure that it won't be tracked in ByteBuffAllocator's
   * reservoir, if both RPC and LRUBlockCache release the block, then it can be garbage collected by
   * JVM, so need a retain here.
```

```
  private Cacheable asReferencedHeapBlock(Cacheable buf) {
    if (buf instanceof HFileBlock) {
      HFileBlock blk = ((HFileBlock) buf);
      if (blk.isSharedMem()) {
        return HFileBlock.deepCloneOnHeap(blk);
      }
    }
    // The block will be referenced by this LRUBlockCache, so should increase its refCnt here.
    return buf.retain();
  }
```

Reply:
@anoopsjohn This is simplified now.